service/history/workflow/mutable_state_impl.go

```diff
@@ -1663,9 +1669,27 @@ func (ms *MutableStateImpl) UpdateActivity(
 
 	ms.pendingActivityInfoIDs[ai.ScheduledEventId] = ai
 	ms.updateActivityInfos[ai.ScheduledEventId] = ai
+	ms.updateActivityInfosUserDataUpdated[ai.ScheduledEventId] = struct{}{}
 	ms.approximateSize += ai.Size() - prev.Size()
 	return nil
 }
+func (ms *MutableStateImpl) UpdateActivityTimerTaskStatus(
```

Comment:
nit: let's add some comment on `UpdateActivity()` saying it ideally should not be used and if used, can only be used when user data needs to be changed. Same for `UpdateUserTimer()`